### ISSUE TITLE
YOLOV11N MODEL CONVERSION TO tensorRT AND INTERIGATION INTO JETSON

### ISSUE THREAD

**Issue by sanskarjain1503** [ultralytics/ultralytics#22074]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

PLEASE HELP ME I WAS DEVELOPED YOLOV11N MODEL ON OWN DATASET USING PRE TRAINED MODEL NOW I HAVE TO CONVERT IT INTO tensorRT and INTRIGATED WITH JETSON ORIN NANO PLEASE GUIDE ME STEP BY STEP WHAT I WILL DO I VISIT IN ULTRALYTICS BUT I HAVE NOT FIND A PROPER THNING THERE SO GUIDE BY STET BY STEP

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sanskarjain1503, thank you for your interest in Ultralytics 🚀! This is an automated assistant response — an Ultralytics engineer will also review your issue and assist you soon.

For getting started and general workflows, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and answers to common questions.

Since your request is about converting a trained YOLO11n model to TensorRT and deploying on Jetson Orin Nano, please share the details below so we can guide you step by step:
- Exact `ultralytics` version and Jetson environment details (JetPack version, CUDA/CUDNN, Python, PyTorch)
- A brief summary of the steps/commands you’ve already tried for ONNX/TensorRT export and any errors/logs you encountered
- Your intended runtime (Python vs. C++), precision targets (FP32/FP16/INT8), and expected input size
- A small snippet of your current export/inference code and console output

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```

## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share the details above, we’ll help outline the exact steps to export YOLO11n and integrate it with TensorRT on your Jetson Orin Nano ✅

**Comment by Y-T-G** [comment]
We have a docs page for NVIDIA Jetson describing all the installation steps and also how to convert to TensorRT:

https://docs.ultralytics.com/guides/nvidia-jetson/#use-tensorrt-on-nvidia-jetson

**Comment by Acuno41** [comment]
Hi everyone.

I also trained my custom model on my PC and export it with ultralytics to tensorrt .engine format. It worked very well on the PC.
Then i copied my model.pt training file to my Jetson Orin NX and tried to do same process. The model exported successfully on Jetson but inference accuracy was very bad. It barely detects couple object on jetson which should be 20-30 on same image on my computer
Can we directly use model.pt which trained on my computer on jetson like that ? or should we do something else ?

Best regards

**Comment by Y-T-G** [comment]
What's the command you used to export?

Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by Acuno41** [comment]
I have done it on my python code like below. Where model_path it the full/path/of/model.pt

```python
    def _convertTensorRT(self, model, model_path):
        # Export the model to TensorRT format
        tensorrt_path = model_path[:-3] + ".engine"
        if not(os.path.exists(tensorrt_path)):
            model.export(format="engine",  # creates '.engine'
                         half=self.params.HALF_PRECISION,
                         dynamic=True,
                         batch=1, 
                         nms=self.params.AGNOSTIC_NMS,
                         device=self.params.DEVICE
                        )
        
        return YOLO(tensorrt_path, task=self.params.TASK)
```

and yolo checks output below

```
Ultralytics 8.3.203 🚀 Python-3.10.12 torch-2.5.0a0+872d972e41.nv24.08 CUDA:0 (Orin, 15656MiB)
Setup complete ✅ (8 CPUs, 15.3 GB RAM, 56.7/455.9 GB disk)

OS                     Linux-5.15.148-tegra-aarch64-with-glibc2.35
Environment            Linux
Python                 3.10.12
Install                pip
Path                   /home/hair-analyzer-d0/.local/lib/python3.10/site-packages/ultralytics
RAM                    15.29 GB
Disk                   56.7/455.9 GB
CPU                    ARMv8 Processor rev 1 (v8l)
CPU count              8
GPU                    Orin, 15656MiB
GPU count              1
CUDA                   12.6

numpy                  ✅ 1.23.5>=1.23.0
matplotlib             ✅ 3.5.1>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 9.0.1>=7.1.2
pyyaml                 ✅ 5.4.1>=5.3.1
requests               ✅ 2.25.1>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.5.0a0+872d972e41.nv24.8>=1.8.0
torch                  ✅ 2.5.0a0+872d972e41.nv24.8!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.20.0a0+afc54f7>=0.9.0
psutil                 ✅ 7.1.0
polars                 ✅ 1.33.1
ultralytics-thop       ✅ 2.0.17>=2.0.0

```

i have the pytorch 2.7 on my pc and 2.5 on jetson. Does this make a effect like that ?

**Comment by Y-T-G** [comment]
Remove `nms` argument

**Comment by Acuno41** [comment]
It changed nothing.
I tried to predict with loading .pt file and infer and set score threshold it detects OK.
But poor results still the same optimized .engine

**Comment by Y-T-G** [comment]
What's the TensorRT version?

**Comment by Acuno41** [comment]
The `pip3 list` TensorRT version is below

```bash
tensorrt                     10.3.0
tensorrt_dispatch            10.3.0
tensorrt_lean                10.3.0
tensorstore                  0.1.74
```
and also output of the `dpkg -l | grep nvinfer` is below

```bash
ii  libnvinfer-bin                               10.3.0.30-1+cuda12.5                        arm64        TensorRT binaries
ii  libnvinfer-dev                               10.3.0.30-1+cuda12.5                        arm64        TensorRT development libraries
ii  libnvinfer-dispatch-dev                      10.3.0.30-1+cuda12.5                        arm64        TensorRT development dispatch runtime libraries
ii  libnvinfer-dispatch10                        10.3.0.30-1+cuda12.5                        arm64        TensorRT dispatch runtime library
ii  libnvinfer-headers-dev                       10.3.0.30-1+cuda12.5                        arm64        TensorRT development headers
ii  libnvinfer-headers-plugin-dev                10.3.0.30-1+cuda12.5                        arm64        TensorRT plugin headers
ii  libnvinfer-lean-dev                          10.3.0.30-1+cuda12.5                        arm64        TensorRT lean runtime libraries
ii  libnvinfer-lean10                            10.3.0.30-1+cuda12.5                        arm64        TensorRT lean runtime library
ii  libnvinfer-plugin-dev                        10.3.0.30-1+cuda12.5                        arm64        TensorRT plugin libraries
ii  libnvinfer-plugin10                          10.3.0.30-1+cuda12.5                        arm64        TensorRT plugin libraries
ii  libnvinfer-samples                           10.3.0.30-1+cuda12.5                        all          TensorRT samples
ii  libnvinfer-vc-plugin-dev                     10.3.0.30-1+cuda12.5                        arm64        TensorRT vc-plugin library
ii  libnvinfer-vc-plugin10                       10.3.0.30-1+cuda12.5                        arm64        TensorRT vc-plugin library
ii  libnvinfer10                                 10.3.0.30-1+cuda12.5                        arm64        TensorRT runtime libraries
ii  python3-libnvinfer                           10.3.0.30-1+cuda12.5                        arm64        Python 3 bindings for TensorRT standard runtime
ii  python3-libnvinfer-dev                       10.3.0.30-1+cuda12.5                        arm64        Python 3 development package for TensorRT standard runtime
ii  python3-libnvinfer-dispatch                  10.3.0.30-1+cuda12.5                        arm64        Python 3 bindings for TensorRT dispatch runtime
ii  python3-libnvinfer-lean                      10.3.0.30-1+cuda12.5                        arm64        Python 3 bindings for TensorRT lean runtime

```

I will try to flash Jetson again to fresh installation. Is it enough to follow [this Ultralytics install guide](https://docs.ultralytics.com/guides/nvidia-jetson/) ?

**Comment by Acuno41** [comment]
I also noticed there is a model output difference between my PC and jetson. When export reads .pt file the outputs like below.

From PC
```bash
PyTorch: starting from 'models/detection/det_yolo11n_v0.1.pt' with input shape (1, 3, 1280, 1280) BCHW and output shape(s) (1, 300, 6) (5.4 MB)
```

From Jetson
```bash
PyTorch: starting from 'models/detection/det_yolo11n_v0.1.pt' with input shape (1, 3, 1280, 1280) BCHW and output shape(s) (1, 12, 33600) (5.4 MB)
```

Could this be the problem ?

**Comment by glenn-jocher** [comment]
Thanks for the details. The shape difference is expected—your PC export included TensorRT NMS `(1, 300, 6)` while the Jetson export is raw logits `(1, 12, 33600)` where `12 = 4 + nc (8 classes)`. For an apples-to-apples check, please re-export on Jetson without NMS, fix the input size, and compare PT vs TensorRT on the same image with identical settings (mismatched `imgsz` is the most common cause of “missing” detections, especially if you trained at 1280 but infer at 640):

```python
from ultralytics import YOLO

# Re-export on Jetson (no NMS), use your training size if it was 1280
YOLO("model.pt").export(format="engine", imgsz=1280, half=True, dynamic=False)

# Compare on the same device, same thresholds
img = "test.jpg"
YOLO("model.pt").predict(img, imgsz=1280, conf=0.25, iou=0.6, device=0)
YOLO("model.engine").predict(img, imgsz=1280, conf=0.25, iou=0.6, device=0)
```

If results still diverge, run a metric comparison and share the full console logs from both runs along with `yolo checks` from your PC so we can see versions are aligned:
```bash
yolo val model=model.engine data=path/to/data.yaml imgsz=1280 conf=0.25 device=0
yolo val model=model.pt     data=path/to/data.yaml imgsz=1280 conf=0.25 device=0
```

Yes, a clean flash followed by the Jetson setup steps is sufficient; for export options (e.g., `nms`, `imgsz`, `dynamic`, `half`) see the TensorRT export docs at [TensorRT export for YOLO11 models](https://docs.ultralytics.com/integrations/tensorrt/).

**Comment by Acuno41** [comment]
@glenn-jocher thanks for the response. 

I tried your first code spinned suggestion on my PC and Jetson and both worked perfect than my original code.
Seems like i have done something wrong or missing configurations while converting my model to tensorRT.

One last question, can i activate agnostic_nms parameter to engine prediction or model export ? 

Thanks for your support.
I'am very grateful.

**Comment by glenn-jocher** [comment]
Glad it’s working now. Agnostic NMS isn’t configurable inside a TensorRT engine; if you export with `nms=True` it’s class-aware and `agnostic_nms` has no effect. Export without NMS and enable class-agnostic NMS at inference:

```python
from ultralytics import YOLO

# Re-export without integrated NMS
YOLO("model.pt").export(format="engine", imgsz=1280, half=True, nms=False)

# Class-agnostic NMS at runtime
YOLO("model.engine").predict("img.jpg", imgsz=1280, conf=0.25, iou=0.6, agnostic_nms=True)
```

```bash
yolo export model=model.pt format=engine imgsz=1280 half=True nms=False
yolo predict model=model.engine source=img.jpg imgsz=1280 conf=0.25 iou=0.6 agnostic_nms=True
```

For the full list of export options see the model export arguments in the docs at [Model Export with Ultralytics YOLO](https://docs.ultralytics.com/modes/export/).
